cpp/memilio/compartments/compartmentalmodel.h

```diff
@@ -198,9 +165,255 @@ struct CompartmentalModel {
 
     Populations populations{};
     ParameterSet parameters{};
+};
+
+namespace details
+{
+// The following functions are not defined anywhere. Their use is to provide type conversions via their return type.
+
+// Function declaration used to remove OmittedTag from the type list of a tuple.
+// First a list of tuples is generated for each Tag in Tags, where the tuple is either of type tuple<Tag>, or if
+// Tag == OmittedTag, of type tuple<>. This list is then concatenated, effectively removing OmittedTag.
+template <class OmittedTag, class... Tags>
+decltype(std::tuple_cat(std::declval<typename std::conditional<std::is_same<OmittedTag, Tags>::value, std::tuple<>,
+                                                               std::tuple<Tags>>::type>()...))
+    filter_tuple(std::tuple<Tags...>);
+
+// Function declaration used to replace type T by std::tuple.
+template <template <class...> class T, class... Args>
+std::tuple<Args...> as_tuple(T<Args...>);
+
+// Function declaration used to replace std::tuple by type T.
+template <template <class...> class T, class... Args>
+T<Args...> as_index(std::tuple<Args...>);
+
+// Remove all occurrences of OmittedTag from the types in a std::tuple<types...>.
+template <class OmittedTag, class Tuple>
+using filtered_tuple_t = decltype(filter_tuple<OmittedTag>(std::declval<Tuple>()));
+
+// Remove all occurrences of OmittedTag from the types in an Index = IndexTemplate<types...>.
+template <class OmittedTag, template <class...> class IndexTemplate, class Index>
+using filtered_index_t = decltype(as_index<IndexTemplate>(
+    std::declval<filtered_tuple_t<OmittedTag, decltype(as_tuple(std::declval<Index>()))>>()));
+
+} //namespace details
+
+/**
+ * @brief CompartmentalModel with flows. 
+ *
+ * Inherits from CompartmentalModel extending it with a TypeChart (the optional template parameter). 
+ * @see CompartmentalModel
+ *
+ * Instead of directly computing the right-hand-side with get_derivatives, first the flows defined by TypeChart are
+ * computed, and then added to/subtracted from their source/target compartment.
+ */
+template <class Comp, class Pop, class Params, class... Flows>
+struct CompartmentalModel<Comp, Pop, Params, TypeChart<Flows...>> : public CompartmentalModel<Comp, Pop, Params> {
```

Comment:
The idea was to make the new "FlowModel" a specialization of CompartmentalModel, reusing some of its code and allowing it to be used the same way without modifying any other code.
Maybe actually calling it FlowModel (or something similar) makes this less confusing to look at, and it helps making the documentation more precise (no more "CompartmentalModel with Flows").
Other than that, do you think we should use inheritance to avoid code duplication? Some member functions of CompartmentalModel can be reused 1:1.